### PR TITLE
No tags in tree index

### LINKED ISSUE #798: Remove tagged series from tree index
In order to be compatible with graphite, we need to remove the tagged series from the tree index. If there are series like `some.id.of.a.metric.1;tag1=value1` and `some.id.of.a.metric.2;tag2=value2` then a call to `/metrics/find?query=some.id.of.a.metric.*` should result in `[]`, while current it would result in the two metrics with tags. 
After that change, such tagged metrics will only be queriable via tag based queries.

The easiest way to do that should be to just not add tagged series into the tree index anymore.

It is important that index pruning still works for tagged series, even though they won't be in the tree index anymore.

### PR DESCRIPTION
**Pull request by replay**
As described in #798, tagged series should not be added into the tree index anymore.

This also means that we'll now need separate methods to delete by tag expressions, as described here: http://graphite.readthedocs.io/en/latest/tags.html#removing-series-from-the-tagdb

Furthermore, pruning needs to be fixed so it also takes the tag index into account.

Fixes #798

### CODE REVIEW

**Comment by woodsaj** [comment]
Is it possible to make this behaviour optional via a config setting?  I see some value in being able to have tagged series in the index tree.
eg, this allows users to just start appending tags to existing series and then use the `byTag` functions in their queries. for example,  users could just update their existing collectd config and add a "Postfix" setting to their write_graphite config.  
```
Postfix ";tag1=val1;tag2=val2"
```

**Comment by replay** [comment]
@woodsaj I'm not sure I understand. 
With this patch and also without, once a user starts sending metrics with tags, then those metrics will be queriable via the `byTag` functions as long as tag-support is enabled. So that scenario you're describing should already work.
The difference that this patch makes is that once the metrics have tags, then they will not be queriable via `<name>;<tags>...` anymore because they'll not be in the tree index anymore.

**Comment by woodsaj** [comment]
The issue i see is that with this change, once you start sending tags with a series you can no longer use the query editor to explore the data you have (metric names).  If your series are mostly in the old format with only a couple of tags, that is a problem.

So you would go from being able to navigate the tree from `collectd -> collectd.hostA -> collectd.hostA.df -> collectd.hostA.df.root -> etc...`  to now having to know you are that you are looking for a series with a `name=collectd.hostA.df.root.percent_bytes.free`

**Comment by Dieterbe** [comment]
I think awood's problem consists of 2 pieces:
1) metric navigation/browsing: there's nothing preventing us from providing the ability to navigate/browse the series tree when you use tags, it's just that it would be in a different place than the current autocomplete editor. IOW this aspect would be solved by adding an autocomplete/tree-thing to the tag value editor when the tag key is `name`. 
2) keeping dashboards/panels healthy when people add tags in the way that AJ described. that would no longer work. people would have to update their metric queries from the traditional non-tagged query, to a new tagged query where they fill in the old pattern in the value field for a tag with key `name`. this is essentially the concern as i also raised in #798 and i'm also still not convinced this is the right way to go

**Comment by replay** [comment]
Ok, I understand what you mean now.  I think at least for graphite running in front of MT it should be no problem to just forward `/find` results that include tagged series, it would just be kind of inconsistent with how the tagdb implementation in graphite behaves, but if we make it optional we could disable it by default for consistency with graphite's implementation.

**Comment by replay** [comment]
I think if we do that then `DeleteByTag()` might potentially get quite expensive if somebody deletes by a tag&value combination that results in a large number of series. Because first we'll need to look up all the `MetricDefinition`s by tag query, then for each of their path's we'll need to do a find on the tree index and clean up all the branches with the `delete()` method. But i guess that's just expected then

**Comment by replay** [comment]
@Dieterbe  i think this would be ready for another review

**Comment by Dieterbe** [comment]
since @woodsaj has some specific thoughts on what the implementation should look like, I think he should review first.

File: idx/memory/memory.go [hunk 1]
```diff
@@ -274,7 +269,7 @@ func (m *MemoryIdx) Load(defs []schema.MetricDefinition) int {
 
 		m.add(def)
 
-		if tagSupport {
+		if TagSupport {
```
Comment (woodsaj) [inline idx/memory/memory.go]: should this be `if TagSupport && len(def.Tags) > 0 {` ?

File: idx/memory/memory.go [hunk 2]
```diff
@@ -1095,49 +1139,96 @@ func (m *MemoryIdx) delete(orgId int, n *Node, deleteEmptyParents, deleteChildre
...
+
+			toPruneUntagged[def.OrgId][n.Path] = struct{}{}
+		} else {
+			id, err := idx.NewMetricIDFromString(def.Id)
+			if err != nil {
+				log.Error(3, "memory-idx: corrupt index. ID format %s seems to be invalid", def.Id)
+				continue DEFS
 			}
+
+			toPruneTagged[def.OrgId][id] = struct{}{}
```
Comment (woodsaj) [inline idx/memory/memory.go]: This doesn't take into account there being multiple metricDefs with the same tagset due to the interval of the metric changing (or schema.MetricDefinition.SetId() changing).   When there are multiple metricDefs with the same tagSet we need to keep all of them as it means that series is still in use.

This is why when deleting from the tree, we check to make sure the leaf node has no other metricDefs that are still being updated.

An easy way to solve this would be to add a `DefByTagSet map[string][]*idx.Archive` to the MemoryIdx struct.
where the key of the map is just the series' tags sorted and concatenated together.  Then when pruning, we use the old tree search for non-tagged series, and also iterate over the DefByTagSet   map to find series where all metricDefs are stale.

Comment (replay) [reply idx/memory/memory.go]: right, i wrongly assumed that unlike the tree index there's going to be only one metric def per combo. I forgot about the interval. will fix that

File: idx/memory/memory.go (hunk 1 above)
Comment (replay) [reply idx/memory/memory.go]: actually not, it should just be `if TagSupport` because `indexTags` also indexes the `name` as tag. But while looking at this i noticed that i'll need to fix the other call-site of `indexTags()` accordingly

Comment (replay) [reply idx/memory/memory.go]: fixed

File: idx/memory/memory.go (hunk 2 above)
Comment (replay) [reply idx/memory/memory.go]: Reading your proposal with the `DefByTagSet`, I can already hear @Dieterbe telling us that there are too many strings in the indexes and we don't want more because of gc overhead for references. 
Unfortunately the only good alternative I can think of would involve creating a `TagQuery` for each of the series that are candidates for pruning, but I think that's probably going to cause too much over head.

How about this though:
Instead of using those string indexes for `DefByTagSet` as you describe, we could use a key that is an md5 hash (of static size, not involving references). Then, when the pruning wants to check if for a pruning-candidate there are other series with the same name + tags combo, it could get its md5 hash and lookup the list of other defs with the same md5 hash from `DefByTagSet`. If there are hash collisions it is not a problem because the pruning process could simply ignore those with a matching MD5 but where name + tags don't match the pruning candidate it is currently checking for. 
That way we get all the good things: 1) not many references and 2) it's fast

Or am I making that unnecessarily complicated?

Comment (woodsaj) [reply idx/memory/memory.go]: > Or am I making that unnecessarily complicated?

Yes.   Just use a string.  Strings are only problematic when you start manipulating them as changing a string requires an allocation and copy.
Even if you used a hash of the tagset, that hash value would need to be a []byte (there would be way too many collisions for an integer hash like Murmur or FNV-1a).  A []byte has more overheads then a string. https://syslog.ravelin.com/byte-vs-string-in-go-d645b67ca7ff

Comment (replay) [reply idx/memory/memory.go]: Ok, no problem. 
Btw I didn't mean to use `[]byte`, since an md5 hash is always 128 bits we could use something like a struct of two `uint64` to store it. But I'll use strings anyway

Comment (Dieterbe) [reply idx/memory/memory.go]: > Strings are only problematic when you start manipulating them as changing a string requires an allocation and copy.

yes and no. avoiding allocations is a good thing when we can, especially in hot code, because then you can see speedups and GC is triggered less frequently. (most of our allocations currently happens in kafka-mdm data decoding, every single metric packet ingested triggers a bunch of string allocations)
However the problem replay was referring to is amount of live pointers. the more live pointers (pointers that stick around for minutes, hours, days or weeks) you have, the more the GC has to work each time it runs. and this has an averse effect on performance, slowing down our goroutines to an extent sometimes significantly more than what the stop-the-world reported would lead one to believe. (see #172)

this is why storing stuff in pointerless structures such as integers or fixed-size arrays is beneficial, in general. maybe not so much if the flipside is that you have to calculate checksums. i haven't gone through this particular code yet.

Comment (replay) [reply idx/memory/memory.go]: I think this should solve the described problem then: https://github.com/grafana/metrictank/pull/806/commits/cc0c3d10c46cbc9abb8c1c7c113577caa1861c4a
If in the future we want to get rid of the strings as map index and replace it with something else, like statically sized MD5s or whatever, then that would be really easy.

Comment (replay) [reply idx/memory/memory.go]: @woodsaj that would be ready for review again if you have time

File: idx/idx.go [hunk 3]
```diff
@@ -168,4 +168,8 @@ type MetricIndex interface {
 	// If the fourth parameter is > 0 then only those metrics of which the LastUpdate
 	// time is >= the from timestamp will be included.
 	TagDetails(int, string, string, int64) (map[string]uint64, error)
+
+	// DeleteTagged deletes the specified series from the tag index and also the
+	// DefById index.
+	DeleteTagged(int, []string) ([]Archive, error)
```
Comment (Dieterbe) [inline idx/idx.go]: now that things are separated out, it's due time for us to document in `MemoryIdx` struct, which attributes are for what use case. it seems some properties are solely for tree index, some are for just tagged index, and some (just DefById ?) are for both. this can be documented/organized better.

File: api/models/node.go [hunk 4]
```diff
@@ -122,6 +122,19 @@ func (t IndexAutoCompleteTagValues) Trace(span opentracing.Span) {
 func (i IndexAutoCompleteTagValues) TraceDebug(span opentracing.Span) {
 }
 
+type IndexTagDelSeries struct {
+	OrgId int      `json:"orgId" binding:"Required"`
+	Paths []string `json:"path" form:"path"`
```
Comment (Dieterbe) [inline api/models/node.go]: interesting that is called "paths". afaik the term "path" is not commonly used yet. but i see it's the same on http://graphite.readthedocs.io/en/latest/tags.html
it looks like these values are used for the values for name tags, so why not just call this name tags ( @DanCech question for you ) :) anyway not worth holding back this PR for my nitpickery but it just caught my eye.

File: idx/memory/memory.go [hunk 5]
```diff
@@ -103,6 +103,29 @@ func (t *TagIndex) delTagId(name, value string, id idx.MetricID) {
 	}
 }
 
+// <name>;<key>=<name> -> Set of references to schema.MetricDefinition
+type defByTagSet map[string]map[*schema.MetricDefinition]struct{}
```
Comment (Dieterbe) [inline idx/memory/memory.go]: why `*schema.MetricDefinition` and not `schema.MetricDefinition` ? would save a lot of pointers.

File: idx/memory/memory.go [hunk 6]
```diff
@@ -220,25 +246,20 @@ func (m *MemoryIdx) indexTags(def *schema.MetricDefinition) {
...
 // to it from the tag index. It assumes a lock is already held.
-func (m *MemoryIdx) deindexTags(def *schema.MetricDefinition) {
-	tags, ok := m.tags[def.OrgId]
-	if !ok {
-		corruptIndex.Inc()
-		log.Error(3, "memory-idx: corrupt index. ID %q can't be removed. no tag index for org %d", def.Id, def.OrgId)
-		return
-	}
-
+func (m *MemoryIdx) deindexTags(tags TagIndex, def *schema.MetricDefinition) bool {
```
Comment (Dieterbe) [inline idx/memory/memory.go]: please document what the bool means

File: idx/memory/tag_query_test.go [hunk 7]
```diff
@@ -483,55 +487,6 @@ func TestGetByTag(t *testing.T) {
 	}
 }
 
-func TestDeleteTaggedSeries(t *testing.T) {
```
Comment (Dieterbe) [inline idx/memory/tag_query_test.go]: why is this removed? is it replaced by something else? if so, what?

File: idx/memory/memory_test.go [hunk 8]
```diff
@@ -40,29 +40,51 @@ func getRandomString(n int, alphabets ...byte) string {
...
 		data[i].SetId()
 	}
+
 	return data
 }
 
+// testWithAndWithoutTagSupport calls a test with all combinations of
+// the settings TagSupport and tagsInTree. In some cases those settings can
+// affect the logic quite a lot, so we need to test all combinations
+func testWithAndWithoutTagSupport(t *testing.T, f func(*testing.T)) {
```
Comment (Dieterbe) [inline idx/memory/memory_test.go]: this is a cool idea, but the mentioned tagsInTree parameter does not exist

Comment (Dieterbe) [reply idx/memory/memory_test.go]: wouldn't it be more meaningful though, if the called functions actually did more with tags?
e.g. `testWithAndWithoutTagSupport(t, testPrune)`  testPrune doesn't have any tagged data, so the tagsupport enabling won't do much.  so the tag support aspect of this is a bit misleading. or should we just document that the purpose here is testing solely non-tag-related stuff in context of an enabled and disabled index?

(maybe out of scope for this PR?)

File: idx/memory/memory.go (hunk 5 above)
Comment (replay) [reply idx/memory/memory.go]: If we copy the MDs then we'd need to make sure that all updates to MetricDefinitions are made to all instances of each. F.e. if we update `LastUpdate` we'd need to make sure that the update is done to all instances of that MetricDefinition, which seems to be complicated and prone to errors.

File: idx/memory/tag_query_test.go (hunk 7 above)
Comment (replay) [reply idx/memory/tag_query_test.go]: The interface to delete tagged series has changed. Previously `Delete()` could be used to delete tagged series, while now those methods have been separated into `Delete()` and `DeleteTagged()`. There are other tests for `DeleteTagged()`, so this one is not necessary anymore.

File: idx/memory/memory_test.go (hunk 8 above)
Comment (replay) [reply idx/memory/memory_test.go]: sorry, that comment is old, the `tagsInTree` parameter did exist but then we decided to remove it.

Comment (replay) [reply idx/memory/memory_test.go]: I think even if a function shouldn't do anything related to tagged series, it's still a good idea to test it with both settings, since doing so is basically free and it just gives a little more assurance that everything works as expected.

Comment (replay) [reply idx/memory/memory_test.go]: https://github.com/grafana/metrictank/pull/806/commits/ad7491ae4bcc2aa656882397f3f93ef484717e0e

File: api/models/node.go (hunk 4 above)
Comment (replay) [reply api/models/node.go]: i think the term name already has an implied meaning of `name` wihout any tags, because we also have a virtual tag called `name` which has that content.
So we need an alternative word to describe `name;<tags>...` and because `nameWithTags` path might be better

File: idx/memory/memory.go (hunk 6 above)
Comment (replay) [reply idx/memory/memory.go]: done: https://github.com/grafana/metrictank/pull/806/commits/8021face0dafa51e4084b172cda1810579f2401a

File: idx/idx.go (hunk 3 above)
Comment (replay) [reply idx/idx.go]: while doing that, i realized that `defByTagSet` needs to be keyed by `orgId`, so I did that, plus added comments to the properties of `MemoryIdx`: https://github.com/grafana/metrictank/pull/806/commits/f319903099d35645bed7899596289a22b2c2dc16

Comment (Dieterbe) [reply idx/idx.go]: I think you misunderstood what i meant.  I meant we need to comment the use case for each attribute (for tag index, for hierarchy index, or for both). there was no need to describe the internals of the structures again since you already described that where the structures are defined, which is the right thing to do.

internals of stuff should be documented where the stuff is defined, not where they are used.

I changed this via 20a4ba2797e1e1ffd9e41f98b08e9caa5d1d7748 please let me know what you think or if you see any mistakes

Comment (replay) [reply idx/idx.go]: yep, that works
